OpenAPI ValidatorをCIに追加

a5ob7r opened the pull request:
# 概要

OpenAPI Validatorを利用してAPI仕様書を検査する。

https://github.com/IBM/openapi-validator

a5ob7r committed:
Add OpenAPI Validator to CI jobs

a5ob7r committed:
Use a cache by hand

a5ob7r committed:
Add a path

a5ob7r committed:
Fix the if statement

a5ob7r commented:
他のAPI仕様書関連のPRがマージされたら、本PRにてリンターにより指摘されたエラーを修正するなどして本PRをマージするつもりである。

epaew commented:
（PR 内容とは関係ない話）

GitHub の Pull Request を Draft に戻すには、ここをクリックすればよいです。  
右側に縦に並んでる中の revewers セクション？ね

![スクリーンショット 2023-06-12 12 14 44](https://github.com/sls-training/2023-API-training-server/assets/6658740/b438dbf4-bc1c-42ff-b7a0-5f292bc51b0e)

epaew commented on .github/workflows/ci.yml:
```diff
@@ -67,6 +67,32 @@ jobs:
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: actions/setup-node@v3
```
https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#caching-packages-data

`actions/setup-node` はそれ単体で node_modules をキャッシュする機能を持ってるので、  
素直に `package.json` をリポジトリに置くパターンであれば、`actions/cache` と組み合わせる必要はないのよねえ。。  
個人的には今の使い方は「うーん…わざわざ面倒なことしてない？」ってなってる。

---
いっそ「一番シンプルな書き方」に振り切って、
```sh
npx ibm-openapi-validator@1.1.1 openapi.yaml
```
に切り替えちゃうのもいいんじゃないかなと思いました。

https://docs.npmjs.com/cli/v9/commands/npx

（just an idea であり、「変えてくれ」ではありません）

tk0miya replied:
へー、それ知らなかった
https://github.com/actions/setup-node/tree/main#caching-global-packages-data

の

>Note: The action does not cache node_modules

を読んで、なんだか面倒くさいぞ…? と思考停止してましたw

a5ob7r replied:
インストールしたいツールはリンター一つのみでしたし、色々と書いてみたいお年頃だったので、手作業でキャッシュ管理をしていましたが、設定を管理する手間を考慮して、node-setup組み込みのキャッシュ機構を利用することにしました。

ただ、キャッシュする必要ないだろうと思うくらいの時間で依存関係のインストールが完了してますね。そこはなんというか、礼儀としてキャッシュしておくことにしました。
https://github.com/sls-training/2023-API-training-server/actions/runs/5353851066/jobs/9710184543

Taku3939 commented on .github/workflows/ci.yml:
```diff
@@ -67,6 +67,32 @@ jobs:
+        run: npm install -g ibm-openapi-validator@1.1.1
+
+      - name: Add a path for OpenAPI Validator
+        run: echo ~/.node_modules/bin >> "$GITHUB_PATH"
```
修正しなくても良いと思うが、$GITHUB_PATHのクオーテーションってなくても良さそう？

tk0miya replied:
ファイル名にスペースが入っていると意図せぬ問題が起きるので、ファイル名には常にスペースを付けておくのが安心ですね。今回は問題ないのだけど、習慣としてやっておくに越したことはないです。

※ 意識的に問題があるところ、無いところを見極めて使い分けができるのであれば、端折ってもよいのですが、我々は愚かなので判断ミスが起きても安心な方に倒しておくのが安全です。

Taku3939 replied:
> ファイル名には常にスペースを付けておくのが安心ですね。

スペースをつけるのですか...？

確かに、スペース入ってることを考慮するとそうなるのか...。今まで環境変数にわざわざクォーテーションつけるの見なかったので勉強なりました。

a5ob7r replied:
僕の中のクォーテーション警察の指示でクォーテーションしました。

でも、見てみたらドキュメントのほうではやっていないですね。

```shell
echo "$HOME/.local/bin" >> $GITHUB_PATH
```

https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-system-path

a5ob7r replied:
余談ですけど、Bash等のシェルだと field splitting が発生するんですけど、Zshでは発生しないんですよね。 field splitting というのは、変数展開後に、展開された値を再度スペースでトークンに分割する処理のことです。

```shell
var='foo bar'

# Bash
: >> $var
: >> foo bar
# fooとbarという2つのファイルが作成される。

# Zsh
: >> $var
: >> 'foo bar'
# 'foo bar'という1つのファイルが作成される。
```

a5ob7r replied:
そういえば、なんでこれでパスを追加できるのかよくわからないんですよね。どういう理屈なんだろう。

epaew replied:
> スペースをつけるのですか...？

（多分間違えただけだと思います。）

Taku3939 replied:
気になってちょっと調べてみたらそれっぽい記事があったので貼っときます
https://qiita.com/xtetsuji/items/381dc17241bda548045d

tk0miya replied:
あれ、打ち合わせしながら書いたのでめっちゃ typo してる… ファイル名には常に"クォート" をつけておくのが安心、です…

a5ob7r committed:
Fix a field name

a5ob7r committed:
Switch to the spectral's ruleset

This is more general than the IBM Cloud Validation Ruleset, which is by
default and more specific for IBM Cloud.

a5ob7r committed:
Fix a mismatch between the schema and the example

a5ob7r committed:
Add package.json and the lock file

The lock file is used by the setup-node action to manage caches in CI.

a5ob7r committed:
Switch to setup-node native caching

Though it uses actions/cache internally, we no need to consider it and
can be simplify the configuration.

a5ob7r commented:
放置していましたが、いくつか修正してマージ可能な状態になりました。レビューをしていただけるとありがたいです。

- package.jsonおよびpackage-lock.jsonを追加してnode-setup組み込みのキャッシュ機構を利用
    - CIのログを見ると、キャッシュ不要ではと思うくらいあっという間に依存関係のインストールが完了するけど、礼儀としてキャッシュしておく。
    - https://github.com/sls-training/2023-API-training-server/actions/runs/5353851066/jobs/9710184543
- 利用するルールセットをより汎用的なものに変更
    - 以前はIBM Cloud用のルールセットを利用していたが、こちらは当然サービス固有のルールが含まれていたりして、少し使いにくいので、より汎用的なOpenAPI用のルールセットに変更した。
    - https://github.com/IBM/openapi-validator/blob/main/docs/ibm-cloud-rules.md
    - https://github.com/stoplightio/spectral/blob/develop/docs/reference/openapi-rules.md

a5ob7r commented:
あと、リンターに指摘されたいくつかのエラーを修正しました。

tk0miya commented on .github/workflows/ci.yml:
```diff
@@ -67,6 +67,23 @@ jobs:
+          cache: npm
+
+      - name: Install dev dependencies
+        run: npm ci
```
`npm ci` しらなかった。へー
https://qiita.com/mstssk/items/8759c71f328cab802670

epaew replied:
少し古い NPM であれば  `i (install)` `ci (clean install)` の使い分けが必須でしたが、最近の NPM なら、`i` でもちゃんと `package-lock.json` を見てくれるようになってるはずなので、使い分けが必要な場面はかなり減りましたね

a5ob7r replied:
node-setupのREADMEに記載されていたので、そのまま拝借しました。ciサブコマンドについて知らなかったのと、CIの設定をしていたために脳内がCI一色だったので、ci = CIかと思っていました。いや、どんなコマンドだよ。
https://github.com/actions/setup-node#caching-global-packages-data

tk0miya replied:
continuous integration じゃなくて clean install なのね。そんなもんだろうと思って納得して調べもしなかったw

epaew replied:
continuous integration なのであれば `npm install && npm run test` が流れてほしい（）

まあ実際 CI 向きなので、この場面では「`i` よりも `ci` のほうがふさわしい」には変わりないという理解です。

- https://docs.npmjs.com/cli/v9/commands/npm-install?v=true
- https://docs.npmjs.com/cli/v9/commands/npm-ci?v=true

tk0miya reviewed (approved):
すでに +1 済みだけど、改めて心の +1 投げとく

a5ob7r commented:
Approve数が2つになったのでマージします。ありがとうございました。
